### PR TITLE
feat: add cache store for proxy

### PR DESCRIPTION
**Pull request by gaius-qi**
## Description
This pull request introduces several changes to the Dragonfly client, focusing on adding LRU cache functionality to the proxy server and making various improvements and additions across different modules.

### Cache Implementation and Integration:

* Added a new `Cache` struct in `dragonfly-client/src/proxy/cache.rs` for storing HTTP responses using the LRU algorithm. This includes methods for creating a new cache, getting content by request, and adding pieces to the cache.
* Integrated the new `Cache` into the `Proxy` struct and its associated methods in `dragonfly-client/src/proxy/mod.rs`. This includes passing the cache to various handler functions and using it to manage piece content. [[1]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR75-R77) [[2]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR110) [[3]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR171) [[4]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cL178-R184) [[5]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR208) [[6]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR225) [[7]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR237) [[8]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR254) [[9]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR266) [[10]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR279) [[11]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR288) [[12]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR301) [[13]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR311) [[14]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR325) [[15]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cL343-R367) [[16]](diffhunk://#diff-6421011d65f9faa2a693ea46e321cc00e07b77ef8792a1b795d1db98d9e5ec4cR380)

### Configuration Enhancements:

* Added a new configuration parameter `cache_capacity` to the `Proxy` struct in `dragonfly-client-config/src/dfdaemon.rs`, including a default value and integration into the `Default` implementation. [[1]](diffhunk://#diff-1d4c7839a6a08bcaef14196c90b20366303b24b2826e1bc9123574a815e7c1ddR1092-R1097) [[2]](diffhunk://#diff-1d4c7839a6a08bcaef14196c90b20366303b24b2826e1bc9123574a815e7c1ddR1112)
* Added a function `default_proxy_cache_capacity` to provide the default cache capacity value.

### Dependency Updates:

* Added `lru = "0.12.5"` to the dependencies in `Cargo.toml` and included it in the workspace dependencies. [[1]](diffhunk://#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R95) [[2]](diffhunk://#diff-651fbc571d827abadd8e14ed21f83bc11941b8733f2d9704b93ae13b4dbfbb31R66)

### Error Handling:

* Added new error variants `Infallible` and `Utf8` to the `DFError` enum in `dragonfly-client-core/src/error/mod.rs` to handle infallible and UTF-8 errors.

### Code Annotations:

* Added `#[inline]` attributes to various functions across multiple files to suggest inlining to the compiler for performance improvements. [[1]](diffhunk://#diff-09bca31ba146d9bc116819606a37c3f4073946a1e2af40382b3c41e440869e6eR444) [[2]](diffhunk://#diff-58e1a6463adbab70df687ead673404e28a954e0a3e955683356e63eda00dcfe9R832) [[3]](diffhunk://#diff-0facc381d7000044e4fade29b772456b0234ce8f1fa844d8f487bafa375bb6f1R60) [[4]](diffhunk://#diff-0facc381d7000044e4fade29b772456b0234ce8f1fa844d8f487bafa375bb6f1R71) [[5]](diffhunk://#diff-0facc381d7000044e4fade29b772456b0234ce8f1fa844d8f487bafa375bb6f1R121) [[6]](diffhunk://#diff-0facc381d7000044e4fade29b772456b0234ce8f1fa844d8f487bafa375bb6f1R151)

These changes collectively enhance the caching capabilities of the Dragonfly client, improve configuration flexibility, update dependencies, and refine error handling and performance.

### CODE REVIEW

File: dragonfly-client/src/proxy/cache.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                .piece
+                .calculate_interested(piece_length, content_length, range)?;
+
+        // Calculate the content capacity based on the interested pieces and push the content into
+        // the bytes.
+        let content_capacity = interested_pieces.len() * piece_length as usize;
+        let mut content = bytes::BytesMut::with_capacity(content_capacity);
+        for interested_piece in interested_pieces {
+            let piece_id = self.task.piece.id(&task_id, interested_piece.number);
+            let Some(piece_content) = self.get_piece(&piece_id) else {
```
Comment (CormickKneey) [inline dragonfly-client/src/proxy/cache.rs:84]: Should we add metric here to evaluate cache miss rate for future optimization?

**Review by chlins — APPROVED** [review]
lgtm

File: dragonfly-client/src/proxy/mod.rs [hunk 2]
```diff
@@ -595,6 +636,25 @@ async fn proxy_by_dfdaemon(
...
+        Ok(None) => {
+            debug!("cache miss");
+        }
+        Ok(Some(content)) => {
+            debug!("cache hit");
+            let body_boxed = Full::new(content).map_err(ClientError::from).boxed();
+            return Ok(Response::new(body_boxed));
+        }
+        Err(err) => {
+            error!("get content from cache failed: {}", err);
```
Comment (yxxhero) [inline dragonfly-client/src/proxy/mod.rs:650]: why not ignore this error? could we continue if has err when get cache?

File: dragonfly-client/src/proxy/cache.rs (hunk 1 above)
Comment (gaius-qi) [reply dragonfly-client/src/proxy/cache.rs:82]: Added.

File: dragonfly-client/src/proxy/mod.rs (hunk 2 above)
Comment (gaius-qi) [reply dragonfly-client/src/proxy/mod.rs:666]: Rocksdb error needs to fail fast.

**Pull request merged**